File: from_others/k-mst/source/framework/src/Instance.cpp
#include "Instance.h"

Instance::Instance( string file )
{
	ifstream ifs( file.c_str() );
	if( ifs.fail() ) {
		cerr << "could not open input file " << file << endl;
		exit( -1 );
	}

	cout << "Reading instance from file " << file << endl;

	ifs >> n_nodes >> n_edges;
	cout << "Number of nodes: " << n_nodes << endl;
	cout << "Number of edges: " << n_edges << endl;

	edges.resize( n_edges );
	incidentEdges.resize( n_nodes );

	u_int id;
	while( ifs >> id ) {
		ifs >> edges[id].v1 >> edges[id].v2 >> edges[id].weight;
		incidentEdges[edges[id].v1].push_back( id );
		incidentEdges[edges[id].v2].push_back( id );
	}
	ifs.close();

	cout << "Incidency list:" << endl;
	for( u_int v = 0; v < n_nodes; v++ ) {
		cout << v << ": ";
		for( list<u_int>::iterator li = incidentEdges[v].begin();
				li != incidentEdges[v].end(); li++ ) {
			cout << "(" << edges[*li].v1 << "," << edges[*li].v2 << "), ";
		}
		cout << endl;
	}
}

File: java_workspace/kMinSpanningTreeCPLEX/src/kMinSpanningTreeCPLEX/Instance.java
package kMinSpanningTreeCPLEX;

import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class Instance {
	/**
	 * array of edges
	 */
	private Edge[] edges;

	/**
	 * incident edges denoted by index in list edges
	 */
	private List<List<Integer>> incidentEdges;

	/**
	 * number of nodes
	 */
	private int numberOfNodes;

	/**
	 * number of edges
	 */
	private int numberOfEdges;

	public Instance(String file) {
		Path inputFile = Paths.get(file);
		if (!Files.exists(inputFile)) {
			System.err.println("could not open input file " + file);
			System.exit(-1);
		}

		System.out.println("Reading instance from file " + file);
		List<String> lines;
		try {
			lines = Files.readAllLines(inputFile, Charset.defaultCharset());

			// TODO add back edge for every edge with same weight
			
			numberOfNodes = Integer.parseInt(lines.remove(0));
			numberOfEdges = Integer.parseInt(lines.remove(0));

			edges = new Edge[numberOfEdges];

			// initialize incident edges array
			incidentEdges = new ArrayList<List<Integer>>();
			for (int i = 0; i < numberOfNodes; i++) {
				incidentEdges.add(new ArrayList<Integer>());
			}

			for (String line : lines) {
				if (!line.isEmpty()) {
					String[] values = line.split(" ");
					int id = Integer.parseInt(values[0]);
					int v1 = Integer.parseInt(values[1]);
					int v2 = Integer.parseInt(values[2]);
					int weight = Integer.parseInt(values[3]);

					// add new edge
					edges[id] = new Edge(v1, v2, weight);
					incidentEdges.get(edges[id].getV1()).add(id);
					incidentEdges.get(edges[id].getV2()).add(id);
				}
			}

			System.out.println("Incidency list:");
			for (int v = 0; v < numberOfNodes; v++) {
				System.out.print(v + ": ");

				for (int e : incidentEdges.get(v)) {
					System.out.print("(" + edges[e].getV1() + ","
							+ edges[e].getV2() + "), ");
				}
				System.out.println();
			}

		} catch (IOException e) {
			System.err.println(e.getMessage());
			System.exit(-1);
		}

	}

	public int getNumberOfNodes() {
		return numberOfNodes;
	}

	public int getNumberOfEdges() {
		return numberOfEdges;
	}

	public Edge getEdge(int i) {
		return edges[i];
	}

	public List<Integer> getIncidentEdges(int node) {
		return incidentEdges.get(node);
	}
}
